### PR TITLE
Master

### LINKED ISSUE #12: Implement the changes in Exercise #3 (ch 6)
Add a factorial operator: use a suffix ! operator to represent “factorial.”
For example, the expression 7! means 7 6 5 4 3 2 1. Make ! bind tighter
than * and /; that is, 7*8! means 7*(8!) rather than (7*8)!. Begin by
modifying the grammar to account for a higher-level operator. To agree
with the standard mathematical definition of factorial, let 0! evaluate to 1.
Hint: The calculator functions deal with doubles, but factorial is defined
only for ints, so just for x!, assign the x to an int and calculate the
factorial of that int

### LINKED ISSUE #20: David-Code Review PR#19
Review Aaron's implementation of the factorial feature and '{' & '}' support

### PR DESCRIPTION
**Pull request by awsumpter**
Updated exercise 2&3 code

### CODE REVIEW

File: Calculator/Calculator.cpp [hunk 1]
```diff
@@ -43,35 +42,38 @@ Token_stream::Token_stream()
...
     if (full) error("putback() into a full buffer");
-    buffer = t; // copy t to buffer
-    full = true; // buffer is now full
+    buffer = t; 
+    full = true; //Buffer cannot be overwritten
 }
 
 //------------------------------------------------------------------------------
+// Evaluates the buffer and assembles and returns Token(ch) or Token(ch,val) depending 
+// on character input
```
Comment (davrdic) [inline Calculator/Calculator.cpp:52]: good stuff

File: Calculator/Calculator.cpp [hunk 2]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
 
 //------------------------------------------------------------------------------
 
-Token_stream ts;       // provides get() and putback()
+Token_stream ts;       
 
 //------------------------------------------------------------------------------
 
-double expression();  // read and evaluate a Expression
+double expression();
```
Comment (davrdic) [inline Calculator/Calculator.cpp:92]: why are these here? lol don't worry about commenting these now just consider if you should or not

File: Calculator/Calculator.cpp [hunk 3]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
...
 
 //------------------------------------------------------------------------------
 
-double expression();  // read and evaluate a Expression
+double expression();  
 
 //------------------------------------------------------------------------------
+// Gets next token from the token stream and evaluates expression grouped within
+// '{}' or '()' as well as single number expressions and returns the value 
+// for the grouped expression or single number
```
Comment (davrdic) [inline Calculator/Calculator.cpp:97]: good stuff

File: Calculator/Calculator.cpp [hunk 4]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
...
+            }
+            return f;
+        }
+        default:
+            ts.putback(t);
+            return left;
+        }
+    }
+}
+
```
Comment (davrdic) [inline Calculator/Calculator.cpp:153]: impressive really. even though the guts of how you calculate your factorial is truly fucked up...where you have it placed in the design as well as how it is handled is exactly like mine. no need to make any changes now. we can go over my final answer and you can get some ideas for future reference.

File: Calculator/Calculator.cpp [hunk 5]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
...
     default:
         error("primary expected");
     }
 }
 
 //------------------------------------------------------------------------------
+// Gets next token from token stream and calculates a factorial denoted by '!' and
+// returns the value for the factorial. All tokens other than '!' are putback into the
+// token stream
+
```
Comment (davrdic) [inline Calculator/Calculator.cpp:130]: good stuff

File: Calculator/Calculator.cpp [hunk 6]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
...
+        }
+    }
+}
+
+//------------------------------------------------------------------------------
+// Gets next token from token stream and evaluates a term. Calculates multiplication or division
+// of a term by a factorial or primary and returns the result. All tokens other than '*' and '/'
+// are put back into the token stream
 
 double term()
```
Comment (davrdic) [inline Calculator/Calculator.cpp:159]: good stuff

File: Calculator/Calculator.cpp [hunk 7]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
...
             return left;
         }
     }
 }
 
 
 //------------------------------------------------------------------------------
+// Gets next token from the token stream and evaluates expressions. Calculates 
+// addition and subtraction of terms within the expression. All tokens other than
+// '+' and '-' are returned to the token stream
```
Comment (davrdic) [inline Calculator/Calculator.cpp:190]: good stuff

File: Calculator/Calculator.cpp [hunk 8]
```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
...
+            ts.putback(t); 
+            return left; 
         }
     }
 }
 
 //------------------------------------------------------------------------------
 
+
 int main()
```
Comment (davrdic) [inline Calculator/Calculator.cpp:215]: no overview for main? poor main... lol don't worry about it for now

**Review by davrdic — APPROVED** [review]
looks good!i left comments on what you should baseline your comments off of. go ahead and merge. well done on the factorial

File: Calculator/Calculator.cpp (hunk 8 above)
Comment (awsumpter) [reply Calculator/Calculator.cpp:215]: yeah. i was worried about over commenting for some reason so i tried to stick to mainly functions that were doing something, but i can definitely add comment for main as we go along lol
